python/ray/util/sgd/v2/trainer.py

```diff
@@ -226,8 +234,92 @@ def to_tune_trainable(
         Returns:
             A Trainable that can directly be passed into ``tune.run()``.
         """
+        return _create_tune_trainable(train_func, self._backend,
+                                      self._num_workers, self._use_gpu,
+                                      self._resources_per_worker)
+
 
-        def trainable_func(config: Dict[str, Any]) -> T:
-            pass
+class _TuneTrainer(Trainer):
+    def start_training(
+            self,
+            train_func: Union[Callable[[], T], Callable[[Dict[str, Any]], T]],
+            config: Optional[Dict[str, Any]] = None) -> None:
+
+        # from ray.util import inspect_serializability
+        # inspect_serializability(train_func, name="train_func_2")
+
+        train_func = self._get_train_func(train_func, config)
+        self._executor.start_training(train_func)
+
+    def fetch_next_report(self) -> Optional[List[Dict]]:
+        return self._executor.fetch_next_result()
+
+
+class _SgdTrainable(DistributedTrainable):
+    _function = None
+    _backend = None
+    _num_workers = None
+    _use_gpu = None
+    _resources_per_worker = None
+
+    _started = False
+    _finished = False
+
+    def setup(self, config: Dict):
+        # Create a new Trainer with the same configuration.
+        self._trainer = _TuneTrainer(self._backend, self._num_workers,
+                                     self._use_gpu, self._resources_per_worker)
+        # TODO(matt): use placement group resource allocation
+        self._trainer.start()  # Move this to step?
+        self._config = config
+
+    def step(self) -> Dict:
+        if self._finished:
+            raise RuntimeError("Training has already finished.")
+        if not self._started:
+            self._trainer.start_training(self._function, self._config)
```

Comment:
There is a pickling issue caused by line 280. Calling `inspect_serializability` on `self._function` at the start of `step` will result in a failure, but calling the same check at the start of `setup` will not.
```python
        from ray.util import inspect_serializability
        import time
        import sys

        inspect_serializability(self._function, name="test_func")
        time.sleep(2)
        sys.exit(0)
```
```
(pid=44303) ================================================================================
(pid=44303) Checking Serializability of <bound method train_func of <ray.util.sgd.v2.trainer._WrappedSgdTrainable object at 0x7fc6685e3df0>>
(pid=44303) ================================================================================
(pid=44303) !!! FAIL serialization: cannot pickle '_thread.lock' object
(pid=44303)     Serializing '__func__' <function train_func at 0x7fc6685e0820>...
(pid=44303) WARNING: Did not find non-serializable object in <bound method train_func of <ray.util.sgd.v2.trainer._WrappedSgdTrainable object at 0x7fc6685e3df0>>. This may be an oversight.
(pid=44303) ================================================================================
(pid=44303) Variable: 
(pid=44303) 
(pid=44303) 	FailTuple(test_func [obj=<bound method train_func of <ray.util.sgd.v2.trainer._WrappedSgdTrainable object at 0x7fc6685e3df0>>, parent=None])
(pid=44303) 
(pid=44303) was found to be non-serializable. There may be multiple other undetected variables that were non-serializable. 
(pid=44303) Consider either removing the instantiation/imports of these variables or moving the instantiation into the scope of the function/class. 
(pid=44303) If you have any suggestions on how to improve this error message, please reach out to the Ray developers on github.com/ray-project/ray/issues/
(pid=44303) ================================================================================
```
cc @richardliaw  @amogkam

Reply:
Fixed this by passing `self.__class__._function` instead.